services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenRevocationEndpoint.java

```diff
@@ -0,0 +1,211 @@
+import org.keycloak.protocol.oidc.utils.AuthorizeClientUtil;
+import org.keycloak.representations.RefreshToken;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.managers.UserSessionCrossDCManager;
+import org.keycloak.services.managers.UserSessionManager;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.util.TokenUtil;
+
+/**
+ * @author <a href="mailto:yoshiyuki.tabata.jy@hitachi.com">Yoshiyuki Tabata</a>
+ */
+public class TokenRevocationEndpoint {
+    private static final String PARAM_TOKEN = "token";
+
+    @Context
+    private KeycloakSession session;
+
+    @Context
+    private HttpRequest request;
+
+    @Context
+    private HttpHeaders headers;
+
+    @Context
+    private ClientConnection clientConnection;
+
+    private MultivaluedMap<String, String> formParams;
+    private ClientModel client;
+    private RealmModel realm;
+    private EventBuilder event;
+    private Cors cors;
+    private RefreshToken token;
+    private UserModel user;
+
+    public TokenRevocationEndpoint(RealmModel realm, EventBuilder event) {
+        this.realm = realm;
+        this.event = event;
+    }
+
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    public Response revoke() {
+        event.event(EventType.REVOKE_GRANT);
+
+        cors = Cors.add(request).auth().allowedMethods("POST").auth().exposedHeaders(Cors.ACCESS_CONTROL_ALLOW_METHODS);
+
+        checkSsl();
+        checkRealm();
+        checkClient();
+
+        formParams = request.getDecodedFormParameters();
+
+        checkToken();
+        checkIssuedFor();
+
+        checkUser();
+        revokeClient();
+
+        event.detail(Details.REVOKED_CLIENT, client.getClientId()).success();
+
+        return cors.builder(Response.ok()).build();
+    }
+
+    private void checkSsl() {
+        if (!session.getContext().getUri().getBaseUri().getScheme().equals("https")
+            && realm.getSslRequired().isRequired(clientConnection)) {
+            throw new CorsErrorResponseException(cors.allowAllOrigins(), OAuthErrorException.INVALID_REQUEST, "HTTPS required",
+                Response.Status.FORBIDDEN);
+        }
+    }
+
+    private void checkRealm() {
+        if (!realm.isEnabled()) {
+            throw new CorsErrorResponseException(cors.allowAllOrigins(), "access_denied", "Realm not enabled",
+                Response.Status.FORBIDDEN);
+        }
+    }
+
+    private void checkClient() {
+        AuthorizeClientUtil.ClientAuthResult clientAuth = AuthorizeClientUtil.authorizeClient(session, event);
+        client = clientAuth.getClient();
+
+        event.client(client);
+
+        cors.allowedOrigins(session, client);
+
+        if (client.isBearerOnly()) {
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_CLIENT, "Bearer-only not allowed",
+                Response.Status.BAD_REQUEST);
+        }
+    }
+
+    private void checkToken() {
```

Comment:
It looks like you are missing validations based on the `token_type_hint`. As well as what is the default value we expect if none was provided.

I understand this parameter is OPTIONAL and we can infer the token type from the token itself (As we are using JWT as the format). But not validate this parameter can cause confusion given that we are supposed to accept only a specific set of values.

Reply:
@pedroigor 
> But not validate this parameter can cause confusion given that we are supposed to accept only a specific set of values.

We are not supposed to accept only a specific set of values. RFC7009 defined two values "access_token" and "refresh_token", and also says that other values for this parameter can be defined. And RFC 7009 also says like [1], so they can be summarized as follows.

- token_type_hint == "refresh_token":
We will support this and if "token" is the refresh token, the server can revoke it. But if "token" is not the refresh token, return an error.

- token_type_hint == "access_token":
We will not support this. But if "token" is the refresh token, the server can revoke it.

- token_type_hint == "invalid_token_type":
The same as "access_token", we will not support this, but if "token" is the refresh token, the server can revoke it.

- token_type_hint == null:
The same as "access_token", we will not support this, but if "token" is the refresh token, the server can revoke it.

According to the above, there is no need to check "token_type_hint".


[1] If the server is unable to locate the token using the given hint, it MUST extend its search across all of its supported token types. An authorization server MAY ignore this parameter, particularly if it is able to detect the token type automatically. An invalid token type hint value is ignored by the authorization server and does not influence the revocation response.

Reply:
@y-tabata I missed that part. But still confusing to me. IMO, the AS could just have a default hint (e.g.: refresh) if none was provided and discard right away any value not supported by the server.

My point here is that now we only support revocation using refresh tokens, so we could just fail fast when attempting to pass any other hint.